lightning/src/onion_message/router.rs

```diff
@@ -0,0 +1,248 @@
+			err: "Failed to find a path to the given destination".to_owned(),
+			action: ErrorAction::IgnoreError,
+		}),
+		Err(Error::FinalHopConversion) => {
+			debug_assert!(invalid_final_hop_pk.is_some());
+			Err(LightningError {
+				err: format!("Public key {:?} is invalid", invalid_final_hop_pk),
+				action: ErrorAction::IgnoreAndLog(Level::Trace)
+			})
+		}
+	}
+}
+
+#[derive(Debug, PartialEq)]
+/// Errored running `dijkstra`.
+enum Error {
+	/// No path exists to the destination.
+	PathNotFound,
+	/// Converting the processing hop type to the final hop type failed, see `dijkstra`'s
+	/// `convert_final_hop` parameter.
+	FinalHopConversion,
+}
+
+//  Heavily adapted from https://github.com/samueltardieu/pathfinding/blob/master/src/directed/dijkstra.rs
+//  TODO: how2credit the repo (is that necessary?)?
+/// Run Dijkstra's from `start` until `found_target` indicates that we've found the destination.
+/// `successor_callback` must invoke the callback that it is provided on each of a given node's
+/// next-hop peers. `convert_final_hop` may be used to convert an intermediate processing hop type
+/// (`N`) to a final path hop type (`H`).
+fn dijkstra<N, H, FN, FS, FC>(start: N, successor_callback: &mut FN, found_target: FS,
+	convert_final_hop: &mut FC) -> Result<Vec<H>, Error>
+	where N: Eq + Hash + Copy + Ord,
+	      FN: FnMut(N, &mut FnMut(&N, u64)),
+	      FS: Fn(&N) -> bool,
+	      FC: FnMut(&N) -> Result<H, ()>,
+{
+	let mut to_see = BinaryHeap::new();
+	to_see.push((start, 0));
+	let mut parents: HashMap<N, (N, u64)> = HashMap::new();
+	parents.insert(start, (start, 0));
+
+	let mut target_reached = None;
+	while let Some((node, cost)) = to_see.pop() {
+		let &(_, c) = parents.get(&node).unwrap();
+		if found_target(&node) {
+			target_reached = Some(node);
+			break;
+		}
+		// We may have inserted a node several times into the binary heap if we found a better way to
+		// access it. Ensure that we are currently dealing with the best path and discard the others.
+		if cost > c {
+			continue;
+		}
+		successor_callback(node, &mut |successor, move_cost| {
+			let new_cost = cost + move_cost;
+			match parents.entry(*successor) {
+				hash_map::Entry::Vacant(e) => {
+					e.insert((node, new_cost));
+					to_see.push((*successor, new_cost));
+				}
+				hash_map::Entry::Occupied(mut e) => {
+					if e.get().1 > new_cost {
+						e.insert((node, new_cost));
+						to_see.push((e.get().0, new_cost));
+					}
+				}
+			}
+		});
+	}
+
+	match target_reached {
+		Some(t) => reverse_path(parents, t, convert_final_hop).map_err(|()| Error::FinalHopConversion),
+		None => Err(Error::PathNotFound)
+	}
+}
+
+// Errors if `convert_path_hop` fails.
+fn reverse_path<N, H, FC>(parents: HashMap<N, (N, u64)>, start: N, convert_path_hop: &mut FC) -> Result<Vec<H>, ()>
+	where N: Eq + Hash + Copy + Ord,
+				FC: FnMut(&N) -> Result<H, ()>,
+{
+	let mut path = vec![convert_path_hop(&start)?];
+	let mut curr = start;
```

Comment:
IIUC, this `start` is actually the target hop. So renaming the variable accordingly could spare future confusion.